src/module_3/module3_statistical_learning.md

```diff
@@ -0,0 +1,1368 @@
+                                      Set for reproducible results. Defaults to None.
+
+    Returns:
+        tuple: A tuple containing three DataFrames: (train_df, validation_df, test_df)
+    """
+
+    # Get unique order IDs
+    unique_order_ids = df['order_id'].unique()
+
+    # Split order IDs into train and test groups 
+    train_order_ids, test_order_ids = train_test_split(unique_order_ids, 
+                                                       test_size=test_size, 
+                                                       random_state=random_state)
+
+    # Optionally, split train order IDs further into train and validation groups
+    if validation_size > 0:
+        train_order_ids, validation_order_ids = train_test_split(train_order_ids,
+                                                                 test_size=validation_size / (1 - test_size),
+                                                                 random_state=random_state)
+
+    # Create split masks
+    train_mask = df['order_id'].isin(train_order_ids)
+    validation_mask = df['order_id'].isin(validation_order_ids)
+    test_mask = df['order_id'].isin(test_order_ids)
+
+    # Split the DataFrame based on masks
+    train_df = df[train_mask]
+    validation_df = df[validation_mask]
+    test_df = df[test_mask]
+
+    return train_df, validation_df, test_df
+
+train_df_0, validation_df_0, test_df_0 = split_by_order_id(df_orders)
+```
+
+### Create a baseline model
+
+
+```python
+train_df = train_df_0.copy()
+validation_df = validation_df_0.copy()
+
+### Label Encoder
+encoders = {}
+for col in categorical_cols:
+    le = LabelEncoder()
+    train_df[col] = le.fit_transform(train_df[col])
+    encoders[col] = le
+
+# Save the encoding to reuse it in validation and test
+with open(data_path + 'label_encoders.pkl', 'wb') as f:
+    pickle.dump(encoders, f)
+
+# ------------------ Model Training ------------------
+    
+X_train = train_df.drop(label_col, axis=1)
+y_train = train_df[label_col]
+
+# Fit the LogisticRegression model
+log_reg = LogisticRegression() 
+log_reg.fit(X_train, y_train) 
+
+# ------------------ Classification ------------------
+
+# Encode test data
+with open(data_path + 'label_encoders.pkl', 'rb') as f:
+    encoders = pickle.load(f)
+
+for col in categorical_cols: 
+    validation_df[col] = encoders[col].transform(validation_df[col])
+
+X_val = validation_df.drop(label_col, axis=1)
+y_val = validation_df[label_col]
+
+# Make predictions
+y_pred = log_reg.predict(X_val) 
+
+```
+
+
+```python
+def model_performance_metrics(y_test, y_pred):
+    accuracy = accuracy_score(y_test, y_pred)
+    conf_matrix = confusion_matrix(y_test, y_pred, labels=[1, 0])
+    precision = precision_score(y_test, y_pred, labels=[1, 0])
+    recall = recall_score(y_test, y_pred, labels=[1, 0])
+    f1 = f1_score(y_test, y_pred)
+    roc_auc = roc_auc_score(y_test, y_pred)
+
+    print(f'Accuracy: {accuracy}')
```

Comment:
accuracy / precision and recall do not make sense if they don't come with a threshold and therefore I really discourage them as "raw" metrics of a classifier. Instead use ROC AUC or PR AUC